Move SAML connection validation after auth checks

camscale opened the pull request:
Move the validation for SAML connectors being inserted or updated (Upsert)
from before the role check for Insert/Update to after it. The validation
can perform an HTTP GET request if the `entity_descriptor_url` field is
set. This should at least require that the user have permission to
Upsert a SAML OIDC connector.

camscale committed:
Move SAML connection validation after auth checks

Move the validation for SAML connectors being inserted or updated (Upsert)
from before the role check for Insert/Update to after it. The validation
can perform an HTTP GET request if the `entity_descriptor_url` field is
set. This should at least require that the user have permission to
Upsert a SAML OIDC connector.

codingllama commented on lib/auth/saml.go:
```diff
@@ -46,6 +46,9 @@ var ErrSAMLNoRoles = trace.AccessDenied("No roles mapped from claims. The mappin
 
 // UpsertSAMLConnector creates or updates a SAML connector.
 func (a *Server) UpsertSAMLConnector(ctx context.Context, connector types.SAMLConnector) error {
+	if err := services.ValidateSAMLConnector(connector, a); err != nil {
```
Is this covered by tests? Could we get a simple test case for the bug so we don't regress?

camscale replied:
I'll see how simple I can make it. I'll need to test that an Upsert with invalid data for an unauthorized user returns an authorization error and not a validation failure error. I will also stand up a test http server and point the entity_descriptor_url to that and ensure it does not get hit in the case of an authorization failure.

I'll also check the code path for OIDC and Github connectors - I have checked that they dont have this same exact issue, but I believe they can have URLs to (maybe that's just OIDC though), so a similar test should be done for them.

codingllama replied:
Thanks for making the effort.

zmb3 commented:
We might as well fix the unbounded http.Get while we're here.

reedloden commented:
@camscale See https://github.com/gravitational/SecOps/issues/212 and https://github.com/gravitational/SecOps/issues/211 as well, if that helps.

camscale committed:
Create roles in tests referenced in SAML connectors

Ensure that roles exist that are referenced in SAML connectors, as the
roles in SAML connectors are validated as the connector is inserted or
updated. The validation has moved to auth.Server (from grpcserver) so
that is now required in tests that operate against auth.Server.

codingllama commented on lib/services/saml.go:
```diff
@@ -266,7 +270,7 @@ func GetSAMLServiceProvider(sc types.SAMLConnector, clock clockwork.Clock) (*sam
 
 // UnmarshalSAMLConnector unmarshals the SAMLConnector resource from JSON.
-func UnmarshalSAMLConnector(bytes []byte, opts ...MarshalOption) (types.SAMLConnector, error) {
+func UnmarshalSAMLConnector(ctx context.Context, bytes []byte, opts ...MarshalOption) (types.SAMLConnector, error) {
```
Interesting how the context suddenly exposes a bunch of methods that, naively, I wouldn't expect to depend on networking.

camscale replied:
Yeah, as what started as a small little fix starts to sprawl, I'm wondering if we're even doing this right. Every time it validates, it fetches the entity descriptor if there is a URL there, and updates the entity descriptor field with the result. But I'm not sure that gets stored every time (likely not), so if the results of the GET change, things might get a little weird with hard to reproduce results. But this code has been there for 5 years so I guess it hasn't caused any noticable problems.

codingllama commented on lib/services/saml_test.go:
```diff
@@ -38,9 +39,10 @@ func TestParseFromMetadata(t *testing.T) {
+	oc, err := UnmarshalSAMLConnector(ctx, raw.Raw)
 	require.NoError(t, err)
-	err = ValidateSAMLConnector(oc, nil)
+	err = ValidateSAMLConnector(ctx, oc, nil)
```
Doesn't Unmarshal call Validate? Could we drop it from the test?

camscale replied:
Yep, I've dropped that

camscale commented on lib/auth/auth_test.go:
```diff
@@ -1104,6 +1104,11 @@ func TestSAMLConnectorCRUDEventsEmitted(t *testing.T) {
 	})
 	require.NoError(t, err)
 
+	role, err := types.NewRole("dummy", types.RoleSpecV5{})
```
The role does need to be registered due to the changes I made. It used to be that validation was done in the grpc server, but has now moved to the auth server. So the tests used to work without the role because the saml connector was not validated completely by the auth server, but now is. Some recent changes added the role checking, but that role checking is not always done because it needs a RoleGetter (auth.Server or interface to) and that is not available in all places.

I'll comment appropriately and deduplicate the strings as you suggest.

camscale commented:
I've put this PR into draft because there are still some things to do (tests for the actual change, and fixing up some breakage from plumbing the context through to the http request). I'll undraft and PTAL when it's ready again.

camscale committed:
pr: Use role GetName() in connector

Use role.GetName() in the connector role mapping config to deduplicate
the name of the role in the test. Add some comments describing the
necessity of the roles being created.

camscale committed:
pr: Add test case for moved validation

Add a test case to test that an access denied error is returned when
upserting an invalid SAML connector instead of a validation failure,
showing that we are not using input from untrusted sources. Also test
that validation is still performed when access is permitted and that a
valid SAML connector can be upserted.

camscale committed:
pr: Remove ValidateSAMLConnector from test

Do not call `ValidateSAMLConnector` after `UnmarshalSAMLConnector` as
the latter already calls the validate function.

camscale committed:
Merge branch 'master' into camh/validate-saml-later

camscale commented:
I've dropped the change to make the http.Get cancelable - I've made a dogs breakfast of my repo multiple times managing/coordinating the OSS & enterprise repos and I still can't get the tests to pass. I clearly have something to learn but I can't figure it out. While I would like to clean up as I move through the code, this cleanup will have to be done after because it is stalling me getting what is a pretty simple fix into the repo.

I have added the test cases for the moved code and addressed the comments related to that. The comments relating to context are no longer relevant for this RP.

Apologies for the force push if it interrupts your review flow. I can't get back my original state.

camscale commented on lib/auth/saml_test.go:
```diff
@@ -227,6 +227,11 @@ func TestPingSAMLWorkaround(t *testing.T) {
 		PrivateKey: fixtures.EncryptionKeyPEM,
 	}
 
+	role, err := types.NewRole("admin", types.RoleSpecV5{})
```
Done

camscale commented on lib/services/local/resource.go:
```diff
@@ -376,10 +376,10 @@ func itemToOIDCConnector(item backend.Item) (types.OIDCConnector, error) {
 // instance of `backend.Item` suitable for storage.
 func itemFromSAMLConnector(connector types.SAMLConnector) (*backend.Item, error) {
-	if err := services.ValidateSAMLConnector(connector, nil); err != nil {
+	if err := services.ValidateSAMLConnector(context.TODO(), connector, nil); err != nil {
```
Dropped context changes from this PR

camscale commented:
PTAL @codingllama

codingllama commented on lib/auth/grpcserver_test.go:
```diff
@@ -2532,3 +2534,103 @@ func TestExport(t *testing.T) {
+				edSvrCalled = true
+				_, err := w.Write([]byte(tc.entityDescriptor))
+				require.NoError(t, err)
+
```
nit:

```suggestion
```

camscale replied:
done

codingllama commented on lib/auth/grpcserver_test.go:
```diff
@@ -2532,3 +2534,103 @@ func TestExport(t *testing.T) {
+			ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+			defer cancel()
+
+			srv := newTestTLSServer(t)
```
Maybe:

```suggestion
			server := newTestTLSServer(t)
```

(or `tlsServer`)

camscale replied:
done

codingllama commented on lib/auth/grpcserver_test.go:
```diff
@@ -2532,3 +2534,103 @@ func TestExport(t *testing.T) {
+
+			srv := newTestTLSServer(t)
+			// Create an http server to serve the entity descriptor url
+			edSvrCalled := false
+			edSvr := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```
Maybe:

```suggestion
			entityServerCalled := false
			entityServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```

camscale replied:
done

codingllama commented on lib/auth/grpcserver_test.go:
```diff
@@ -2532,3 +2534,103 @@ func TestExport(t *testing.T) {
+		allow            types.RoleConditions
+		entityDescriptor string
+		edURLCalled      bool
+		errAs            error
```
Maybe:

```suggestion
		entityDescriptor string
		assertErr        func(error) bool
```

So you can pass in `trace.IsAccessDenied` or `trace.IsBadParameter` directly? (Just be careful with `nil` in the test body.)

camscale replied:
done. no special nil handling required, except for what was already there. `trace.IsBadParameter(nil)` is valid code. Added an explicit `require.Error(t, err)` so the error message is clear if we didn't get an error at all.

codingllama commented on lib/auth/grpcserver_test.go:
```diff
@@ -2532,3 +2534,103 @@ func TestExport(t *testing.T) {
+			err = client.UpsertSAMLConnector(ctx, connector)
+
+			if tc.errAs != nil {
+				require.ErrorAs(t, err, &tc.errAs)
```
```suggestion
				require.ErrorAs(t, err, &tc.errAs, "UpsertSAMLConnector error type mismatch")
```

camscale replied:
done

codingllama reviewed (commented):
Thanks for the fix, Cam. Added a few minor comments.

camscale committed:
pr: Rename vars, use existing error checking fns

Address PR comments, renaming vars to longer names, and use existing
error checking predicates for the error cases.

camscale committed:
Merge branch 'master' into camh/validate-saml-later

camscale commented:
@codingllama comments addressed. PTAL

camscale committed:
Merge branch 'master' into camh/validate-saml-later

camscale committed:
pr: remove stupid test failure

Made a change to ensure the test would fail properly, forgot to remove
it. sigh.
